File: partiql-plan/src/main/java/org/partiql/plan/rel/RelScan.java
package org.partiql.plan.rel;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operator;
import org.partiql.plan.OperatorVisitor;
import org.partiql.plan.rex.Rex;

import java.util.List;

/**
 * Logical scan abstract base class.
 */
public abstract class RelScan extends RelBase {

    /**
     * @return new {@link RelScan} instance
     */
    @NotNull
    public static RelScan create(@NotNull Rex rex) {
        return new Impl(rex);
    }

    /**
     * @return input rex (operand 0)
     */
    @NotNull
    public abstract Rex getRex();

    @NotNull
    @Override
    protected final RelType type() {
        throw new UnsupportedOperationException("Derive type is not implemented");
    }

    @NotNull
    @Override
    protected final List<Operator> operands() {
        Rex c0 = getRex();
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitScan(this, ctx);
    }

    @NotNull
    public abstract RelScan copy(@NotNull Rex rex);

    private static class Impl extends RelScan {

        private final Rex rex;

        private Impl(Rex rex) {
            this.rex = rex;
        }

        @NotNull
        @Override
        public Rex getRex() {
            return rex;
        }

        @NotNull
        @Override
        public RelScan copy(@NotNull Rex rex) {
            return new Impl(rex);
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/rel/RelScan.kt
package org.partiql.plan.rel

import org.partiql.plan.rex.Rex

/**
 * Logical scan corresponding to the clause `FROM <expression> AS <v>`.
 */
public interface RelScan : Rel {

    public fun getInput(): Rex

    override fun getChildren(): Collection<Rel> = emptyList()

    override fun isOrdered(): Boolean = false

    override fun <R, C> accept(visitor: RelVisitor<R, C>, ctx: C): R = visitor.visitScan(this, ctx)
}

/**
 * Default [RelScan] implementation.
 */
internal class RelScanImpl(input: Rex) : RelScan {

    // DO NOT USE FINAL
    private var _input: Rex = input

    override fun getInput(): Rex = _input

    override fun getType(): RelType {
        TODO("Implement getSchema for scan")
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || other !is RelScan) return false
        return _input == other.getInput()
    }

    override fun hashCode(): Int {
        return _input.hashCode()
    }
}
